File: src-tauri/src/conflicts.rs
//! Conflict detection module for Opta.
//!
//! This module provides Tauri commands to detect competitor optimization tools
//! that may conflict with Opta's optimizations. Follows the same pattern as
//! processes.rs by invoking the Python MCP server.

use serde::{Deserialize, Serialize};
use std::process::Command;
use tauri::command;

/// Conflict severity level.
/// - "high": Significant conflict, likely to cause issues
/// - "medium": Partial conflict, may cause some issues
/// - "low": Minor conflict, minimal impact
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ConflictSeverity {
    High,
    Medium,
    Low,
}

/// Information about a detected competitor tool.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConflictInfo {
    /// Internal identifier (e.g., "geforce_experience")
    pub tool_id: String,
    /// Display name (e.g., "NVIDIA GeForce Experience")
    pub name: String,
    /// Brief description of the tool
    pub description: String,
    /// Conflict severity level
    pub severity: String,
    /// Actionable recommendation for the user
    pub recommendation: String,
    /// List of matching process names found
    pub detected_processes: Vec<String>,
}

/// Summary of conflict detection results.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConflictSummary {
    /// Total number of detected conflicts
    pub total_count: u32,
    /// Number of high severity conflicts
    pub high_count: u32,
    /// Number of medium severity conflicts
    pub medium_count: u32,
    /// Number of low severity conflicts
    pub low_count: u32,
    /// Full list of detected conflicts
    pub conflicts: Vec<ConflictInfo>,
}

/// Python script to detect conflicts.
///
/// This inline script invokes the conflicts module from the installed
/// opta_mcp package and returns the conflict summary as JSON.
const DETECT_CONFLICTS_SCRIPT: &str = r#"
import json
import sys

try:
    from opta_mcp.conflicts import get_conflict_summary
    result = get_conflict_summary()
    print(json.dumps(result))
except Exception as e:
    # Return error-safe default response
    print(json.dumps({
        "total_count": 0,
        "high_count": 0,
        "medium_count": 0,
        "low_count": 0,
        "conflicts": [],
        "error": str(e)
    }))
    sys.exit(0)
"#;

/// Detect competitor optimization tools by invoking Python subprocess.
///
/// This command spawns Python to run the conflict detection function from opta_mcp.
/// Returns a summary with detected conflicts and severity counts.
///
/// # Returns
///
/// A `ConflictSummary` containing detected conflicts and counts.
///
/// # Errors
///
/// Returns an error string if Python execution fails or JSON parsing fails.
#[command]
pub async fn detect_conflicts() -> Result<ConflictSummary, String> {
    // Try python3 first, fall back to python
    let python_cmd = if cfg!(target_os = "windows") {
        "python"
    } else {
        "python3"
    };

    let output = Command::new(python_cmd)
        .arg("-c")
        .arg(DETECT_CONFLICTS_SCRIPT)
        .output()
        .map_err(|e| format!("Failed to spawn Python: {}", e))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("Python error: {}", stderr));
    }

    let stdout = String::from_utf8_lossy(&output.stdout);

    // Check if the response contains an error field
    if stdout.contains("\"error\"") {
        let error: serde_json::Value = serde_json::from_str(&stdout)
            .map_err(|e| format!("JSON parse error: {}", e))?;
        if let Some(err_msg) = error.get("error").and_then(|v| v.as_str()) {
            // Log the error but still return the partial result if possible
            eprintln!("Python conflict detection warning: {}", err_msg);
        }
    }

    let result: ConflictSummary = serde_json::from_str(&stdout)
        .map_err(|e| format!("JSON parse error: {} (raw: {})", e, stdout))?;

    Ok(result)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_conflict_info_serialize() {
        let conflict = ConflictInfo {
            tool_id: "geforce_experience".to_string(),
            name: "NVIDIA GeForce Experience".to_string(),
            description: "NVIDIA's game optimization".to_string(),
            severity: "high".to_string(),
            recommendation: "Consider disabling auto-optimization".to_string(),
            detected_processes: vec!["nvcontainer".to_string()],
        };

        let json = serde_json::to_string(&conflict).unwrap();
        assert!(json.contains("geforce_experience"));
        assert!(json.contains("NVIDIA"));
        assert!(json.contains("nvcontainer"));
    }

    #[test]
    fn test_conflict_summary_deserialize() {
        let json = r#"{
            "total_count": 2,
            "high_count": 1,
            "medium_count": 1,
            "low_count": 0,
            "conflicts": [
                {
                    "tool_id": "geforce_experience",
                    "name": "NVIDIA GeForce Experience",
                    "description": "NVIDIA's game optimization",
                    "severity": "high",
                    "recommendation": "Consider disabling",
                    "detected_processes": ["nvcontainer"]
                }
            ]
        }"#;

        let summary: ConflictSummary = serde_json::from_str(json).unwrap();
        assert_eq!(summary.total_count, 2);
        assert_eq!(summary.high_count, 1);
        assert_eq!(summary.conflicts.len(), 1);
        assert_eq!(summary.conflicts[0].tool_id, "geforce_experience");
    }

    #[test]
    fn test_empty_summary_deserialize() {
        let json = r#"{
            "total_count": 0,
            "high_count": 0,
            "medium_count": 0,
            "low_count": 0,
            "conflicts": []
        }"#;

        let summary: ConflictSummary = serde_json::from_str(json).unwrap();
        assert_eq!(summary.total_count, 0);
        assert!(summary.conflicts.is_empty());
    }
}

File: apps/desktop/opta-native/mcp-server/src/opta_mcp/conflicts.py
"""Competitor tool detection for Opta.

Identifies conflicting optimization software (GeForce Experience, Razer Cortex,
MSI Afterburner, OMEN Hub, etc.) that may interfere with Opta's optimizations.

Uses the existing process listing from processes.py to detect running competitors.
Cross-platform support: Windows, macOS, and Linux.
"""

import platform
from typing import Dict, List
from opta_mcp.processes import get_process_list


def get_current_platform() -> str:
    """Get the current platform identifier.

    Returns:
        Platform string: 'windows', 'darwin', or 'linux'
    """
    system = platform.system().lower()
    if system == "darwin":
        return "darwin"
    elif system == "windows":
        return "windows"
    else:
        return "linux"


# Known competitor optimization tools and their characteristics
# Cross-platform with platform-specific entries
COMPETITOR_TOOLS: Dict[str, Dict] = {
    # === Windows-only tools ===
    "geforce_experience": {
        "name": "NVIDIA GeForce Experience",
        "process_names": ["nvcontainer", "nvidia share", "nvidia web helper"],
        "description": "NVIDIA's game optimization and driver updater",
        "conflict_severity": "high",
        "recommendation": "GeForce Experience optimizes the same settings. Consider disabling its auto-optimization.",
        "platforms": ["windows"],
    },
    "razer_cortex": {
        "name": "Razer Cortex",
        "process_names": ["razer cortex", "rzsynapse", "razer central"],
        "description": "Razer's game booster and system optimizer",
        "conflict_severity": "high",
        "recommendation": "Razer Cortex's Game Booster may conflict. Disable it while using Opta.",
        "platforms": ["windows"],
    },
    "msi_afterburner": {
        "name": "MSI Afterburner",
        "process_names": ["msiafterburner", "rtss"],
        "description": "GPU overclocking and monitoring tool",
        "conflict_severity": "medium",
        "recommendation": "Afterburner's OSD and monitoring can coexist, but avoid conflicting GPU settings.",
        "platforms": ["windows"],
    },
    "omen_hub": {
        "name": "HP OMEN Gaming Hub",
        "process_names": ["omen command center", "omen gaming hub"],
        "description": "HP's optimization suite for OMEN devices",
        "conflict_severity": "high",
        "recommendation": "OMEN Hub's performance modes may conflict. Use one optimizer at a time.",
        "platforms": ["windows"],
    },
    "xbox_game_bar": {
        "name": "Xbox Game Bar",
        "process_names": ["gamebar", "gamebarftsvc"],
        "description": "Windows built-in game overlay",
        "conflict_severity": "low",
        "recommendation": "Game Bar's overlay can impact performance. Consider disabling if not needed.",
        "platforms": ["windows"],
    },
    # === macOS-only tools ===
    "cleanmymac": {
        "name": "CleanMyMac",
        "process_names": ["cleanmymac", "macpaw", "cmmhlpr"],
        "description": "macOS system cleaner and optimizer",
        "conflict_severity": "medium",
        "recommendation": "CleanMyMac's optimization features may conflict with Opta. Avoid running simultaneously.",
        "platforms": ["darwin"],
    },
    "parallels_toolbox": {
        "name": "Parallels Toolbox",
        "process_names": ["prl_disp_service", "parallels"],
        "description": "Parallels system utilities and optimization tools",
        "conflict_severity": "low",
        "recommendation": "Parallels Toolbox utilities may run in background. Close if not needed.",
        "platforms": ["darwin"],
    },
    "appcleaner": {
        "name": "AppCleaner",
        "process_names": ["appcleaner"],
        "description": "macOS app uninstaller with cleanup features",
        "conflict_severity": "low",
        "recommendation": "AppCleaner is generally safe but close during optimization sessions.",
        "platforms": ["darwin"],
    },
    # === Linux-only tools ===
    "gamemode": {
        "name": "Feral GameMode",
        "process_names": ["gamemoded", "gamemode"],
        "description": "Linux game performance optimizer daemon",
        "conflict_severity": "high",
        "recommendation": "GameMode provides similar optimizations. Consider disabling GameMode while using Opta.",
        "platforms": ["linux"],
    },
    "mangohud": {
        "name": "MangoHud",
        "process_names": ["mangohud", "mangoappmenu"],
        "description": "Vulkan/OpenGL overlay for monitoring FPS and performance",
        "conflict_severity": "low",
        "recommendation": "MangoHud overlay can coexist. Consider disabling if you see performance issues.",
        "platforms": ["linux"],
    },
    "goverlay": {
        "name": "GOverlay",
        "process_names": ["goverlay"],
        "description": "Linux graphics card management and overlay tool",
        "conflict_severity": "low",
        "recommendation": "GOverlay provides overlays similar to Opta. Close if conflicting visuals occur.",
        "platforms": ["linux"],
    },
    "corectrl": {
        "name": "CoreCtrl",
        "process_names": ["corectrl"],
        "description": "Linux AMD GPU control and overclocking utility",
        "conflict_severity": "medium",
        "recommendation": "CoreCtrl manages GPU settings. Avoid conflicting GPU profiles with Opta.",
        "platforms": ["linux"],
    },
    # === Cross-platform tools ===
    "discord": {
        "name": "Discord (Game Activity)",
        "process_names": ["discord", "discordptb", "discordcanary"],
        "description": "Discord's game activity detection",
        "conflict_severity": "low",
        "recommendation": "Discord's overlay may use resources. Disable overlay if not needed.",
        "platforms": ["windows", "darwin", "linux"],
    },
    "steam_overlay": {
        "name": "Steam Overlay",
        "process_names": ["steamwebhelper", "steam_osx", "steam"],
        "description": "Steam's in-game overlay",
        "conflict_severity": "low",
        "recommendation": "Steam overlay uses minimal resources but can be disabled in Steam settings if needed.",
        "platforms": ["windows", "darwin", "linux"],
    },
}


def detect_running_conflicts() -> List[Dict]:
    """Detect competitor optimization tools that are currently running.

    Uses the process list from processes.py and matches against known
    competitor tools by process name (case-insensitive).
    Only checks tools relevant to the current platform.

    Returns:
        List of detected conflicts, each containing:
        - tool_id: Internal identifier (e.g., "geforce_experience")
        - name: Display name (e.g., "NVIDIA GeForce Experience")
        - description: Brief description of the tool
        - severity: "high", "medium", or "low"
        - recommendation: Actionable advice for the user
        - detected_processes: List of matching process names found
    """
    # Get current process list (already categorized)
    processes = get_process_list()

    # Get current platform for filtering
    current_platform = get_current_platform()

    # Extract all process names (lowercase for comparison)
    running_names = set()
    for proc in processes:
        name = proc.get("name", "").lower()
        if name:
            running_names.add(name)

    detected_conflicts = []

    # Check each competitor tool (filtered by platform)
    for tool_id, tool_info in COMPETITOR_TOOLS.items():
        # Skip tools not relevant to this platform
        tool_platforms = tool_info.get("platforms", ["windows", "darwin", "linux"])
        if current_platform not in tool_platforms:
            continue

        detected_processes = []

        for process_pattern in tool_info["process_names"]:
            pattern_lower = process_pattern.lower()
            # Check if any running process contains this pattern
            for running_name in running_names:
                if pattern_lower in running_name:
                    detected_processes.append(running_name)

        # If any processes matched, add to conflicts
        if detected_processes:
            detected_conflicts.append({
                "tool_id": tool_id,
                "name": tool_info["name"],
                "description": tool_info["description"],
                "severity": tool_info["conflict_severity"],
                "recommendation": tool_info["recommendation"],
                "detected_processes": list(set(detected_processes))  # Remove duplicates
            })

    # Sort by severity (high > medium > low)
    severity_order = {"high": 0, "medium": 1, "low": 2}
    detected_conflicts.sort(key=lambda c: severity_order.get(c["severity"], 3))

    return detected_conflicts


def get_conflict_summary() -> Dict:
    """Get a summary of conflict detection results.

    Returns:
        Dict containing:
        - total_count: Total number of detected conflicts
        - high_count: Number of high severity conflicts
        - medium_count: Number of medium severity conflicts
        - low_count: Number of low severity conflicts
        - conflicts: Full list of detected conflicts
    """
    conflicts = detect_running_conflicts()

    high_count = sum(1 for c in conflicts if c["severity"] == "high")
    medium_count = sum(1 for c in conflicts if c["severity"] == "medium")
    low_count = sum(1 for c in conflicts if c["severity"] == "low")

    return {
        "total_count": len(conflicts),
        "high_count": high_count,
        "medium_count": medium_count,
        "low_count": low_count,
        "conflicts": conflicts
    }
